include/faabric/util/snapshot.h

```diff
@@ -73,27 +76,95 @@ class SnapshotMergeRegion
 class SnapshotData
 {
   public:
-    size_t size = 0;
-    uint8_t* data = nullptr;
-    int fd = 0;
-
     SnapshotData() = default;
 
-    std::vector<SnapshotDiff> getDirtyPages();
+    explicit SnapshotData(size_t sizeIn);
+
+    SnapshotData(size_t sizeIn, size_t maxSizeIn);
+
+    explicit SnapshotData(std::span<const uint8_t> dataIn);
+
+    SnapshotData(std::span<const uint8_t> dataIn, size_t maxSizeIn);
+
+    SnapshotData(const SnapshotData&) = delete;
+
+    SnapshotData& operator=(const SnapshotData&) = delete;
+
+    ~SnapshotData();
+
+    void copyInData(std::span<const uint8_t> buffer, uint32_t offset = 0);
+
+    const uint8_t* getDataPtr(uint32_t offset = 0);
+
+    std::vector<uint8_t> getDataCopy();
+
+    std::vector<uint8_t> getDataCopy(uint32_t offset, size_t dataSize);
 
-    std::vector<SnapshotDiff> getChangeDiffs(const uint8_t* updated,
-                                             size_t updatedSize);
+    void mapToMemory(uint8_t* target);
 
     void addMergeRegion(uint32_t offset,
                         size_t length,
                         SnapshotDataType dataType,
                         SnapshotMergeOperation operation,
                         bool overwrite = false);
 
+    void clearMergeRegions();
+
+    std::map<uint32_t, SnapshotMergeRegion> getMergeRegions();
+
+    size_t getQueuedDiffsCount();
+
+    void queueDiffs(const std::vector<SnapshotDiff>& diffs);
+
+    void writeQueuedDiffs();
+
+    size_t getSize();
+
+    size_t getMaxSize();
+
   private:
+    size_t size = 0;
+    size_t maxSize = 0;
+
+    int fd = -1;
+
+    std::shared_mutex snapMx;
+
+    MemoryRegion data = nullptr;
+
+    std::vector<SnapshotDiff> queuedDiffs;
+
     // Note - we care about the order of this map, as we iterate through it
     // in order of offsets
     std::map<uint32_t, SnapshotMergeRegion> mergeRegions;
+
+    uint8_t* validatedOffsetPtr(uint32_t offset);
+
+    void mapToMemory(uint8_t* target, bool shared);
+
+    void writeData(std::span<const uint8_t> buffer, uint32_t offset = 0);
+};
+
+class MemoryView
+{
+  public:
+    // Note - this object is just a view of a section of memory, and does not
+    // own the underlying data
+    explicit MemoryView(std::span<const uint8_t> dataIn);
+
+    MemoryView(const MemoryView&) = delete;
+
+    MemoryView& operator=(const MemoryView&) = delete;
```

Comment:
If the MemoryView is just a view object, not a resource-owner, it should be trivially copyable and moveable (with default rather than deleted constructors) - and not wrapped in `std::shared_ptr`s where used, as it's just a pointer+length pair (span), especially because you made it immutable so there's no risk of something modifying a MemoryView object.